Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -460,3 +462,72 @@ extension PathHierarchy.DisambiguationContainer {
         }))
     }
 }
+
+// MARK: Deserialization
+
+extension PathHierarchy {
+    // This is defined in the main PathHierarchy.swift file to access fileprivate properties and PathHierarchy.Node API without making it internally visible.
+    
+    // This mapping closure exist so that we don't encode ResolvedIdentifier values into the file. They're an implementation detail and they are a not stable across executions.
+    
+    /// Decode a path hierarchy from its file representation.
+    ///
+    /// The caller can use `mapCreatedIdentifiers` when encoding and decoding path hierarchies to associate auxiliary data with a node in the hierarchy.
+    ///
+    /// - Parameters:
+    ///   - fileRepresentation: A file representation to decode.
+    ///   - mapCreatedIdentifiers: A closure that the caller can use to map indices to resolved identifiers.
+    init(
+        _ fileRepresentation: FileRepresentation,
+        mapCreatedIdentifiers: (_ identifiers: [ResolvedIdentifier]) -> Void
+    ) {
+        // Generate new identifiers. While building the path hierarchy, the node numbers map to identifiers via index lookup in this array.
+        var identifiers = [ResolvedIdentifier]()
+        identifiers.reserveCapacity(fileRepresentation.nodes.count)
+        for _ in 0 ..< fileRepresentation.nodes.count {
```

Comment:
Super-Nit: I recently learned you can use `.indices` for this kind of thing which I find a bit more readable.

```suggestion
        for _ in fileRepresentation.nodes.indices {
```